pubnub-core/pubnub-core-impl/src/main/kotlin/com/pubnub/internal/CorePNConfiguration.kt

```diff
@@ -0,0 +1,387 @@
+        var googleAppEngineNetworking = false
+
+        /**
+         * Instructs the SDK to use a proxy configuration when communicating with PubNub servers.
+         *
+         * @see [Proxy]
+         */
+        var proxy: Proxy? = null
+
+        /**
+         * @see [ProxySelector]
+         */
+        var proxySelector: ProxySelector? = null
+
+        /**
+         * @see [Authenticator]
+         */
+        var proxyAuthenticator: Authenticator? = null
+
+        /**
+         * @see [CertificatePinner]
+         */
+        var certificatePinner: CertificatePinner? = null
+
+        /**
+         * Sets a custom [HttpLoggingInterceptor] for logging network traffic.
+         *
+         * @see [HttpLoggingInterceptor]
+         */
+        var httpLoggingInterceptor: HttpLoggingInterceptor? = null
+
+        /**
+         * @see [SSLSocketFactory]
+         */
+        var sslSocketFactory: SSLSocketFactory? = null
+
+        /**
+         * @see [X509ExtendedTrustManager]
+         */
+        var x509ExtendedTrustManager: X509ExtendedTrustManager? = null
+
+        /**
+         * @see [okhttp3.ConnectionSpec]
+         */
+        var connectionSpec: ConnectionSpec? = null
+
+        /**
+         * @see [javax.net.ssl.HostnameVerifier]
+         */
+        var hostnameVerifier: HostnameVerifier? = null
+
+        /**
+         * How many times publishing file message should automatically retry before marking the action as failed
+         *
+         * Defaults to `5`
+         */
+        var fileMessagePublishRetryLimit = 5
+
+        var dedupOnSubscribe = false
+        var maximumMessagesCacheSize = DEFAULT_DEDUPE_SIZE
+
+        val pnsdkSuffixes: ConcurrentMap<String, String> = ConcurrentHashMap(mutableMapOf())
+
+        fun generatePnsdk(version: String): String {
+            val joinedSuffixes = pnsdkSuffixes.toSortedMap().values.joinToString(" ")
+            return "PubNub-Kotlin/$version" + if (joinedSuffixes.isNotBlank()) " $joinedSuffixes" else ""
+        }
+
+        @Suppress("DeprecatedCallableAddReplaceWith")
+        @Deprecated("To be used by components", level = DeprecationLevel.WARNING)
+        fun addPnsdkSuffix(vararg nameToSuffixes: Pair<String, String>) {
+            @Suppress("DEPRECATION")
+            addPnsdkSuffix(nameToSuffixes.toMap())
+        }
+
+        @Deprecated("To be used by components", level = DeprecationLevel.WARNING)
+        fun addPnsdkSuffix(nameToSuffixes: Map<String, String>) = pnsdkSuffixes.putAll(nameToSuffixes)
+
+        /**
+         * Retry configuration for requests.
+         *  Defaults to [RetryConfiguration.None].
+         *
+         *  Use [RetryConfiguration.Linear] to set retry with linear delay interval
+         *  Use [RetryConfiguration.Exponential] to set retry with exponential delay interval
+         *  Delay will vary from provided value by random value.
+         */
+        var retryConfiguration: RetryConfiguration = RetryConfiguration.None
```

Comment:
We don't have retryConfiguration available for user of Java SDK yet. We can add it now or add to todo list that will be done after merge is released. Maybe it would be better to add it before release?

Reply:
is there anything that should be different for the Java version than the current Kotlin implementation?
We should aim for feature parity now so if we can expose the same RetryConfiguration to Java SDK that would be perfect

Reply:
todo: to add this to PNConfiguration.java

Reply:
added